src/AST-Core/RBVariableNode.class.st

```diff
@@ -44,33 +45,93 @@ RBVariableNode class >> named: aName start: aPosition [
 		yourself.
 ]
 
+{ #category : #'instance creation' }
+RBVariableNode class >> named: anIdentifierName start: aPosition withKind: aKindClass [
+	
+	| newOne kind |
+	newOne := self named: anIdentifierName start: aPosition.
+	kind := aKindClass new.
+	kind variable: newOne.
+	newOne kind: kind.
+	
+	^ newOne.
+]
+
+{ #category : #'instance creation' }
+RBVariableNode class >> newArgumentNamed: aName [
+
+	^ self named: aName start: 0 withKind: RBArgumentKind
+]
+
+{ #category : #'instance creation' }
+RBVariableNode class >> newGlobalNamed: aName [
+
+	^ self named: aName start: 0 withKind: RBGlobalKind
+]
+
+{ #category : #'instance creation' }
+RBVariableNode class >> newSelfNode [
+
+	^ self named: 'self' start: 0 withKind: RBSelfKind
+]
+
+{ #category : #'instance creation' }
+RBVariableNode class >> newSuperNode [
+
+	^ self named: 'super' start: 0 withKind: RBSuperKind
+]
+
+{ #category : #'instance creation' }
+RBVariableNode class >> newTemporaryNamed: aName [
+
+	^ self named: aName start: 0 withKind: RBTemporaryKind
+]
+
+{ #category : #'instance creation' }
+RBVariableNode class >> newThisContextNode [
+
+	^ self named: 'thisContext' start: 0 withKind: RBThisContextKind
+]
+
 { #category : #comparing }
 RBVariableNode >> = anObject [ 
 	self == anObject ifTrue: [^true].
-	((anObject isKindOf: self class) or: [self isKindOf: anObject class]) 
-		ifFalse: [^false].
-	^self name = anObject name
+	(anObject class = self class) 
+		ifFalse: [ ^ false ].
+
+	kind isUndefined ifTrue: [ ^ self name = anObject name ].
+	anObject kind isUndefined ifTrue: [ ^ self name = anObject name ].
+
+	^ kind class = anObject kind class 
+		ifTrue: [ self name = anObject name ]
+		ifFalse: [ false ].
+	
 ]
 
 { #category : #visiting }
 RBVariableNode >> acceptVisitor: aProgramNodeVisitor [
-	^ aProgramNodeVisitor visitVariableNode: self
+	^ kind acceptVisitor: aProgramNodeVisitor
 ]
 
 { #category : #converting }
 RBVariableNode >> adaptToSemanticNode [
 	
-	self primitiveChangeClassTo: self binding semanticNodeClass new
+	kind := self binding semanticNodeKind new.
```

Comment:
Why we need another entity? Binding objects are already about those types. And the implementors of #semanticNodeKind  shows this very clearly.

Reply:
We can do the same visiting logic (introduced here) through OCVariable classes (which the binding here). 
I would also rename #binging variable into #variable.

Reply:
Yes, I think we should move the Semantic analysis (same part of it) into the AST package.

The binding then can play exactly the role of the strategy for visiting.

Maybe we could use this PR as the first step in this direction. The refactoring it does for clients would look the same...

This means: we would merge this, and then in a second step do a PR that move the Semantic vars into the AST package, replacing the Kind classes.